roofit/roofitcore/src/TestStatistics/RooUnbinnedL.cxx

```diff
@@ -0,0 +1,185 @@
+   }
+
+   // include the extended maximum likelihood term, if requested
+   if (extended_) {
+      if (apply_weight_squared) {
+
+         // Calculate sum of weights-squared here for extended term
+         Double_t sumW2(0), sumW2carry(0);
+         for (Int_t i = 0; i < data_->numEntries(); i++) {
+            data_->get(i);
+            Double_t y = data_->weightSquared() - sumW2carry;
+            Double_t t = sumW2 + y;
+            sumW2carry = (t - sumW2) - y;
+            sumW2 = t;
+         }
+
+         Double_t expected = pdf_->expectedEvents(data_->get());
+
+         // Adjust calculation of extended term with W^2 weighting: adjust poisson such that
+         // estimate of Nexpected stays at the same value, but has a different variance, rescale
+         // both the observed and expected count of the Poisson with a factor sum[w] / sum[w^2] which is
+         // the effective weight of the Poisson term.
+         // i.e. change Poisson(Nobs = sum[w]| Nexp ) --> Poisson( sum[w] * sum[w] / sum[w^2] | Nexp * sum[w] /
+         // sum[w^2] ) weighted by the effective weight  sum[w^2]/ sum[w] in the likelihood. Since here we compute
+         // the likelihood with the weight square we need to multiply by the square of the effective weight expectedW
+         // = expected * sum[w] / sum[w^2]   : effective expected entries observedW =  sum[w]  * sum[w] / sum[w^2] :
+         // effective observed entries The extended term for the likelihood weighted by the square of the weight will
+         // be then:
+         //  (sum[w^2]/ sum[w] )^2 * expectedW -  (sum[w^2]/ sum[w] )^2 * observedW * log (expectedW)  and this is
+         //  using the previous expressions for expectedW and observedW
+         //  sum[w^2] / sum[w] * expected - sum[w^2] * log (expectedW)
+         //  and since the weights are constants in the likelihood we can use log(expected) instead of log(expectedW)
+
+         Double_t expectedW2 = expected * sumW2 / data_->sumEntries();
+         Double_t extra = expectedW2 - sumW2 * log(expected);
+
+         // Double_t y = pdf->extendedTerm(sumW2, data->get()) - carry;
+
+         Double_t y = extra - carry;
+
+         Double_t t = result + y;
+         carry = (t - result) - y;
+         result = t;
+      } else {
+         Double_t y = pdf_->extendedTerm(data_->sumEntries(), data_->get()) - carry;
+         Double_t t = result + y;
+         carry = (t - result) - y;
+         result = t;
+      }
+   }
+
+   // If part of simultaneous PDF normalize probability over
+   // number of simultaneous PDFs: -sum(log(p/n)) = -sum(log(p)) + N*log(n)
+   if (sim_count_ > 1) {
+      Double_t y = sumWeight * log(1.0 * sim_count_) - carry;
+      Double_t t = result + y;
+      carry = (t - result) - y;
+      result = t;
+   }
+
+   // At the end of the first full calculation, wire the caches
+   if (_first) {
+      _first = false;
+      pdf_->wireAllCaches();
+   }
+
+   eval_carry_ = carry;
+   return result;
+}
```

Comment:
I believe this was all modernised / simplified / sped up in the mean time. This looks like this commit would revert it all to a state as it was 2 years ago. This should be checked.

Reply:
Ok, in the most recent commit, I factored out the calculation parts into static functions inside RooNLLVar. These are now also called from RooUnbinnedL. This can now also do batched computation. I also added a test for this.